### PR TITLE
ETag feature for REST API Store

### PR DESCRIPTION
**Pull request by sineth-neranjana**

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java [hunk 1]
```diff
@@ -1496,4 +1501,59 @@ public boolean isDuplicateContextTemplate(String contextTemplate) throws APIMana
...
+                int tenantId = ServiceReferenceHolder.getInstance().getRealmService().getTenantManager().getTenantId(
+                        apiTenantDomain);
+                registryType = ServiceReferenceHolder.getInstance().getRegistryService().getGovernanceUserRegistry(
+                        CarbonConstants.REGISTRY_ANONNYMOUS_USERNAME, tenantId);
+            } else {
+                registryType = registry;
+            }
+            return definitionFromSwagger20.getAPISwaggerDefinitionTimeStamps(apiIdentifier,registryType);
+        } catch (org.wso2.carbon.user.api.UserStoreException e) {
+            e.printStackTrace();
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java]: add error log. do not print stacktrace

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java [hunk 2]
```diff
@@ -1496,4 +1501,59 @@ public boolean isDuplicateContextTemplate(String contextTemplate) throws APIMana
...
+        } catch (RegistryException e) {
+            e.printStackTrace();
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java]: add error log. do not print stacktrace

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/definitions/APIDefinitionFromSwagger12.java [hunk 3]
```diff
@@ -421,4 +418,9 @@ public String generateAPIDefinition(API api) throws APIManagementException {
 
         return mainAPIJson.toJSONString();
     }
+
+    @Override
+    public Map<String, String> getAPISwaggerDefinitionTimeStamps(APIIdentifier apiIdentifier, Registry registry) throws APIManagementException {
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/definitions/APIDefinitionFromSwagger12.java]: format method.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/ApisApi.java [hunk 4]
```diff
@@ -54,6 +43,11 @@ public Response apisApiIdDocumentsDocumentIdContentGet(@ApiParam(value = "**API
     {
     return delegate.apisApiIdDocumentsDocumentIdContentGet(apiId,documentId,xWSO2Tenant,accept,ifNoneMatch,ifModifiedSince);
     }
+
+    public String apisApiIdDocumentsDocumentIdContentGetGetLastUpdatedTime(String apiId,String documentId,String xWSO2Tenant,String accept,String ifNoneMatch,String ifModifiedSince)
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/ApisApi.java]: format method

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/ApisApi.java [hunk 5]
```diff
@@ -77,6 +71,11 @@ public Response apisApiIdDocumentsDocumentIdGet(@ApiParam(value = "**API ID** co
     {
     return delegate.apisApiIdDocumentsDocumentIdGet(apiId,documentId,xWSO2Tenant,accept,ifNoneMatch,ifModifiedSince);
     }
+
+    public String apisApiIdDocumentsDocumentIdGetGetLastUpdatedTime(String apiId,String documentId,String xWSO2Tenant,String accept,String ifNoneMatch,String ifModifiedSince)
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/ApisApi.java]: format method

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/ApisApi.java [hunk 6]
```diff
@@ -100,6 +99,11 @@ public Response apisApiIdDocumentsGet(@ApiParam(value = "**API ID** consisting o
     {
     return delegate.apisApiIdDocumentsGet(apiId,limit,offset,xWSO2Tenant,accept,ifNoneMatch);
     }
+
+    public String apisApiIdDocumentsGetGetLastUpdatedTime(String apiId,Integer limit,Integer offset,String xWSO2Tenant,String accept,String ifNoneMatch)
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/ApisApi.java]: format method

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/ApisApi.java [hunk 7]
```diff
@@ -122,6 +126,11 @@ public Response apisApiIdGet(@ApiParam(value = "**API ID** consisting of the **U
     {
     return delegate.apisApiIdGet(apiId,accept,ifNoneMatch,ifModifiedSince,xWSO2Tenant);
     }
+
+    public String apisApiIdGetGetLastUpdatedTime(String apiId,String accept,String ifNoneMatch,String ifModifiedSince,String xWSO2Tenant)
+    {
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/ApisApi.java]: format method

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/ApisApi.java [hunk 8]
```diff
@@ -144,6 +153,11 @@ public Response apisApiIdSwaggerGet(@ApiParam(value = "**API ID** consisting of
     {
     return delegate.apisApiIdSwaggerGet(apiId,accept,ifNoneMatch,ifModifiedSince,xWSO2Tenant);
     }
+
+    public String apisApiIdSwaggerGetGetLastUpdatedTime(String apiId,String accept,String ifNoneMatch,String ifModifiedSince,String xWSO2Tenant)
+    {
+        return delegate.apisApiIdSwaggerGetGetLastUpdatedTime(apiId,accept,ifNoneMatch,ifModifiedSince,xWSO2Tenant);
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/ApisApi.java]: format method

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/ApisApi.java [hunk 9]
```diff
@@ -1,25 +1,14 @@
 package org.wso2.carbon.apimgt.rest.api.store;
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/ApisApi.java]: do proper code formatting

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/ApisApiService.java [hunk 10]
```diff
@@ -25,5 +11,14 @@
     public abstract Response apisApiIdThumbnailGet(String apiId,String accept,String ifNoneMatch,String ifModifiedSince);
     public abstract Response apisGenerateSdkPost(String apiId,String language,String xWSO2Tenant);
     public abstract Response apisGet(Integer limit,Integer offset,String xWSO2Tenant,String query,String accept,String ifNoneMatch);
+
+    public abstract String apisApiIdDocumentsDocumentIdContentGetGetLastUpdatedTime(String apiId,String documentId,String xWSO2Tenant,String accept,String ifNoneMatch,String ifModifiedSince);
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/ApisApiService.java]: format

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/ApplicationsApi.java [hunk 11]
```diff
@@ -41,12 +30,17 @@
...
 
-    public Response applicationsApplicationIdDelete(@ApiParam(value = "**Application Identifier** consisting of the UUID of the Application.\n",required=true ) @PathParam("applicationId") String applicationId,
+    public Response applicationsApplicationIdDelete(@ApiParam(value = "Application Identifier consisting of the UUID of the Application.\n",required=true ) @PathParam("applicationId") String applicationId,
     @ApiParam(value = "Validator for conditional requests; based on ETag.\n"  )@HeaderParam("If-Match") String ifMatch,
     @ApiParam(value = "Validator for conditional requests; based on Last Modified header.\n"  )@HeaderParam("If-Unmodified-Since") String ifUnmodifiedSince)
     {
     return delegate.applicationsApplicationIdDelete(applicationId,ifMatch,ifUnmodifiedSince);
     }
+
+    public String applicationsApplicationIdDeleteGetLastUpdatedTime(String applicationId,String ifMatch,String ifUnmodifiedSince)
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/ApplicationsApi.java]: format all methods

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/ApplicationsApiService.java [hunk 12]
```diff
@@ -23,5 +12,12 @@
     public abstract Response applicationsGenerateKeysPost(String applicationId,ApplicationKeyGenerateRequestDTO body,String contentType,String ifMatch,String ifUnmodifiedSince);
     public abstract Response applicationsGet(String groupId,String query,Integer limit,Integer offset,String accept,String ifNoneMatch);
     public abstract Response applicationsPost(ApplicationDTO body,String contentType);
+
+    public abstract String applicationsApplicationIdDeleteGetLastUpdatedTime(String applicationId,String ifMatch,String ifUnmodifiedSince);
+    public abstract String applicationsApplicationIdGetGetLastUpdatedTime(String applicationId,String accept,String ifNoneMatch,String ifModifiedSince);
+    public abstract String applicationsApplicationIdPutGetLastUpdatedTime(String applicationId,ApplicationDTO body,String contentType,String ifMatch,String ifUnmodifiedSince);
+    public abstract String applicationsGenerateKeysPostGetLastUpdatedTime(String applicationId,ApplicationKeyGenerateRequestDTO body,String contentType,String ifMatch,String ifUnmodifiedSince);
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/ApplicationsApiService.java]: format

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/dto/ApplicationKeyGenerateRequestDTO.java [hunk 13]
```diff
@@ -34,7 +33,34 @@
...
+    return lastUpdatedTime;
+  }
+  public void setLastUpdatedTime(String lastUpdatedTime){
+    this.lastUpdatedTime=lastUpdatedTime;
+  }
+
+  /**
+  * gets and sets the createdTime for a ApplicationKeyGenerateRequestDTO
+  **/
+
```
Comment (praminda) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/dto/ApplicationKeyGenerateRequestDTO.java]: New line not required

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/dto/ApplicationKeyGenerateRequestDTO.java [hunk 14]
```diff
@@ -34,7 +33,34 @@
...
+  @JsonIgnore
+  public String getCreatedTime(){
+    return createdTime;
+  }
+  public void setCreatedTime(String createdTime){
```
Comment (praminda) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/dto/ApplicationKeyGenerateRequestDTO.java]: Additional new line required

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/dto/ApplicationListDTO.java [hunk 15]
```diff
@@ -29,7 +28,34 @@
   
   private List<ApplicationInfoDTO> list = new ArrayList<ApplicationInfoDTO>();
 
-  
+
+  private String lastUpdatedTime = null;
+
```
Comment (praminda) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/dto/ApplicationListDTO.java]: Please remove all additional new lines

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/dto/DocumentDTO.java [hunk 16]
```diff
@@ -41,7 +40,34 @@
...
+  private String createdTime = null;
+
+  /**
+  * gets and sets the lastUpdatedTime for DocumentDTO
+  **/
+  @JsonIgnore
+  public String getLastUpdatedTime(){
+    return lastUpdatedTime;
+  }
+  public void setLastUpdatedTime(String lastUpdatedTime){
```
Comment (praminda) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/dto/DocumentDTO.java]: Additional new line required

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/dto/DocumentDTO.java [hunk 17]
```diff
@@ -41,7 +40,34 @@
...
+
+  /**
+  * gets and sets the createdTime for a DocumentDTO
+  **/
+
+  @JsonIgnore
+  public String getCreatedTime(){
+    return createdTime;
+  }
+  public void setCreatedTime(String createdTime){
```
Comment (praminda) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/dto/DocumentDTO.java]: Additional new line required

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/dto/DocumentListDTO.java [hunk 18]
```diff
@@ -29,7 +28,34 @@
...
+  private String createdTime = null;
+
+  /**
+  * gets and sets the lastUpdatedTime for DocumentListDTO
+  **/
+  @JsonIgnore
+  public String getLastUpdatedTime(){
+    return lastUpdatedTime;
+  }
+  public void setLastUpdatedTime(String lastUpdatedTime){
```
Comment (praminda) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/dto/DocumentListDTO.java]: Check all DTO classes. There are places with additional new lines and also some methods are missing preceding empty line before method signature.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/impl/ApplicationsApiServiceImpl.java [hunk 19]
```diff
@@ -323,4 +323,61 @@ public Response applicationsApplicationIdDelete(String applicationId, String ifM
...
+     * @return
+     */
+    @Override
+    public String applicationsApplicationIdGetGetLastUpdatedTime(String applicationId, String accept, String ifNoneMatch, String ifModifiedSince) {
+        return RestAPIStoreUtils.getLastUpdatedTimeByApplicationId(applicationId);
+    }
+
+    /**
+     * get the lastUpdatedTime for an application PUT
+     * @param applicationId
```
Comment (praminda) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store/src/main/java/org/wso2/carbon/apimgt/rest/api/store/impl/ApplicationsApiServiceImpl.java]: There should be a empty line after method description

**Pull request merged**